File: tree/1650_lowest_common_ancestor_of_a_binary_tree_iii/solution.go
package main

// 1650. Lowest Common Ancestor of a Binary Tree III
// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree-iii/
//
// Nodes carry a Parent pointer. Find LCA of p and q without access to root.
// Two-pointer trick (LC 160 style): when hit nil, jump to the other start.

type Node struct {
	Val    int
	Left   *Node
	Right  *Node
	Parent *Node
}

func lowestCommonAncestor(p, q *Node) *Node {
	// TODO
	return nil
}

File: tree/1650_lowest_common_ancestor_of_a_binary_tree_iii/solution.py
# 1650. Lowest Common Ancestor of a Binary Tree III
# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree-iii/
#
# Nodes have a `parent` pointer. Find LCA of p and q (no root given).
# Two-pointer trick (same as LC 160 Intersection of Two Linked Lists):
#   walk up from each; when you hit null, jump to the other start.
#   They meet at LCA after at most depth(p) + depth(q) steps.

class Node:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None

def lowestCommonAncestor(p, q):
    # TODO
    return None
